modules/mod-deep-learning/DeepModel.cpp

```diff
@@ -0,0 +1,199 @@
+/**********************************************************************
+
+   Audacity: A Digital Audio Editor
+   Audacity(R) is copyright (c) 1999-2021 Audacity Team.
+
+   DeepModel.cpp
+   Hugo Flores Garcia
+
+******************************************************************/
+
+#include "DeepModel.h"
+
+#include <torch/script.h>
+#include <torch/torch.h>
+
+#include "FileNames.h"
+#include "WaveTrack.h"
+#include "WaveClip.h"
+
+#include "CodeConversions.h"
+
+// DeepModel Implementation
+
+FilePath DeepModel::DLModelsDir()
+{
+   return FileNames::MkDir( wxFileName( FileNames::DataDir(), wxT("deeplearning-models") ).GetFullPath() );
+}
+
+FilePath DeepModel::BuiltInModulesDir()
+{
+   return FileNames::MkDir( wxFileName( FileNames::BaseDir(), wxT("deeplearning-models") ).GetFullPath() );
+}
+
+void DeepModel::LoadResampler()
+{
+   // load the resampler module
+   std::string resamplerPath = audacity::ToUTF8(wxFileName(BuiltInModulesDir(), wxT("resampler.pt"))
+                                       .GetFullPath());
+   try
+   {
+      mResampler = std::make_unique<torch::jit::script::Module>
+                     (torch::jit::load(resamplerPath, torch::kCPU));
+      mResampler->eval();
+   }
+   catch (const std::exception &e)
+   {
+      Cleanup();
+      throw ModelException(XO("Error an error occurred while loading the resampler"), e.what());
+   }
+}
+
+void DeepModel::Load(const std::string &modelPath)
+{
+   try
+   { 
+      // set mResampler
+      LoadResampler();
+
+      // set mModel
+      mModel = std::make_unique<torch::jit::script::Module>
+                  (torch::jit::load(modelPath, torch::kCPU));
+      mModel->eval();
+      
+      // finally, mark as loaded
+      mLoaded = true;
+   }
+   catch (const std::exception &e)
+   {
+      Cleanup();
+      throw ModelException(XO("Error while loading model"), e.what());
+   }
+}
+
+void DeepModel::Load(std::istream &bytes)
+{
+   try
+   {
+      // load the resampler module
+      LoadResampler();
+
+      // load the model to CPU, as well as the metadata
+      mModel = std::make_unique<torch::jit::script::Module>
+                     (torch::jit::load(bytes, torch::kCPU));
+      mModel->eval();
+      
+      // finally, mark as loaded
+      mLoaded = true;
+   }
+   catch (const std::exception &e)
+   {
+      Cleanup();
+      throw ModelException(XO("Error while loading model"), e.what());
+   }
+}
+
+bool DeepModel::IsLoaded() const
+{
+   return mLoaded; 
+}
+
+void DeepModel::Save(const std::string &modelPath) const
+{
+   if (!mLoaded)
+      throw ModelException(XO("attempted save when no module was loaded."), "");
+
+   mModel->save(modelPath);
+}
+
+void DeepModel::SetCard(ModelCardHolder card)
+{
+   // set the card
+   mCard = card;
```

Comment:
Might `move`

Reply:
hmm, when ActiveModel calls `DeepModel::SetCard`, it still needs to hold on to it's copy of the card. wouldn't moving it leave `card` empty?